add default namespace for apps and managed helm releases

soer3n opened the pull request:
**What this PR does / why we need it**:

This PR adds a condition for the new introduced field `defaultNamespace` in ApplicationDefinitions crd to configure a default namespace for related ApplicationInstallations.
This PR also adds a new field to Kubermaticconfiguration crd to set a default namespace for created ApplicationInstallations resources in user clusters. This change is related to https://github.com/kubermatic/dashboard/issues/6858.

**Which issue(s) this PR fixes**:
<!--optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged-->
Fixes #12578

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/kind chore
-->

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:
<!--
Write your release note. Release notes are being used to generate the changelog:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

**Documentation**:
<!--
Please do one of the following options:
- Add a link to the existing documentation
- Add a link to the kubermatic/docs pull request
- If no documentation change is applicable then add:
  - TBD (documentation will be added later)
  - NONE (no documentation needed for this PR)
-->
```documentation
TBD
```

soer3n committed:
add default namespace for apps and managed helm releases

* add field to kubermaticconfiguration crd to make default applicationinstallation namespace configurable
* add condition to set the default appinstallation namespace
* add condition to set the default release namespace configured in related appdefinition

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
update docs after adding crd field

Signed-off-by: soer3n <srenhenning@googlemail.com>

SimonTheLeg commented on pkg/apis/kubermatic/v1/configuration.go:
```diff
@@ -261,6 +261,8 @@ type KubermaticUserClusterConfiguration struct {
 	Addons KubermaticAddonsConfiguration `json:"addons,omitempty"`
 	// SystemApplications contains configuration for system Applications (such as CNI).
 	SystemApplications SystemApplicationsConfiguration `json:"systemApplications,omitempty"`
+	// DefaultApplications contains configuration for default Applications settings.
```
```suggestion
	// DefaultApplications contains configuration for default Application settings.
```

soer3n replied:
Changed the comment accordingly.

soer3n committed:
fix unit tests and add tests for new crd fields

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
fix linting errors after adding new unit tests

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
move default app namespace to ee version only

* add wrapper files for building the feature for ee only
* add ee unit tests
* remove unit test regarding this from general unit tests

Signed-off-by: soer3n <srenhenning@googlemail.com>

SimonTheLeg commented on pkg/ee/default-application-controller/util.go:
```diff
@@ -0,0 +1,39 @@
+/*
```
nit: I think I would re-name this to "namespace.go", since it better captures its purpose

soer3n replied:
Yes, makes sense. I changed the filename to `namespace.go` and of course also `utils_test.go` to `namespace_test.go`.

SimonTheLeg commented on pkg/ee/default-application-controller/util.go:
```diff
@@ -0,0 +1,39 @@
+	appskubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/apps.kubermatic/v1"
+)
+
+func GetAppNamespace(ctx context.Context, application appskubermaticv1.ApplicationDefinition) appskubermaticv1.AppNamespaceSpec {
```
I think we should use pointers for this. No need to make a copy of the object. TBH I don't know why it was using full structs in the original `generateApplicationInstallation` func, but I think going forward we should conserve memory where we can. This comment also applies to all the other newly added funcs throughout the callstack.
```suggestion
func GetAppNamespace(ctx context.Context, application *appskubermaticv1.ApplicationDefinition) appskubermaticv1.AppNamespaceSpec {
```

SimonTheLeg replied:
note: I definitely want this for the func arg.
For the return parameter, I am more flexible. While I would prefer to have it return a pointer, I can live with it returning a copy. One more thing: It it is converted to a pointer in line 33 we need to make a copy of the object first and then return a pointer to the copied object.

soer3n replied:
I changed both, input and return value, to pointers. The address is dereferenced during the generation of an applicationinstallation after the function was called.

SimonTheLeg commented on pkg/ee/default-application-controller/util_test.go:
```diff
@@ -0,0 +1,153 @@
+
+func init() {
+	utilruntime.Must(clusterv1alpha1.AddToScheme(testScheme))
+}
```
This is not needed afaics. Because we are not doing any schema related actions inside the test or in the tested functions. Specifically GetAppNamespace only extracts info directly from the AppDef and that is handled directly inside Go without the need to every call any Kubernetes libs.

Also then delete the testScheme var in line 39.

soer3n replied:
I removed the not needed lines.

SimonTheLeg commented on pkg/controller/seed-controller-manager/default-application-controller/controller_test.go:
```diff
@@ -389,6 +392,31 @@ func TestReconcile(t *testing.T) {
 			},
 		},
+		{
+			name:                        "scenario 13: should create default application in cluster with ready Cilium system application in configured default applicationinstallation namespace",
+			defaultApplicationNamespace: applicationInstallationNamespace,
+			cluster:                     genCluster(clusterName, defaultDatacenterName, false, ciliumCNISettings),
+			systemAppInstallationValues: map[string]any{"status": "ready"},
```
I think this test is independent of the Cilium and the systemAppInstallationValues (especially since we only mock a Cilium state in there as well) and the state has nothing to do with where the AppInstall is synched

soer3n replied:
There is a [condition](https://github.com/kubermatic/kubermatic/blob/main/pkg/controller/seed-controller-manager/default-application-controller/controller.go#L166-L173) for checking the status of cilium app when the cni is managed via application before generating and deploying the default applications. The condition is only checking status field.

SimonTheLeg commented on pkg/controller/seed-controller-manager/default-application-controller/controller.go:
```diff
@@ -334,19 +336,27 @@ func (r *Reconciler) generateApplicationInstallation(application appskubermaticv
-			Namespace: appskubermaticv1.AppNamespaceSpec{
-				Name: application.Name,
-				// This ensures that the namespace is created in the user cluster, if it doesn't already exist.
-				Create: true,
```
If I see this correctly then getAppNamespace for a non-empty DefaultNamespace is just whatever is in that spec. This in turn means that if DefaultNamespace.Create was false in the original (which I think is default if left empty), would result in create being false and then the namespace will not get created. Or am I missing something?

(in case I am missing nothing): I wonder why the unit test is not detecting this, afaics it should throw an error when the namespace does not exist (unless fakeClientProvider handles this for you somehow) 🤔

soer3n replied:
This was moved to the new function `getAppNamespace`. In [ce](https://github.com/kubermatic/kubermatic/blob/70bdd1cca20da9488d38395e56625e6252b82643/pkg/controller/seed-controller-manager/default-application-controller/wrappers_ce.go#L27-L32) exactly the same is returned by the function. In [ee](https://github.com/kubermatic/kubermatic/blob/70bdd1cca20da9488d38395e56625e6252b82643/pkg/controller/seed-controller-manager/default-application-controller/wrappers_ee.go#L28-L30) a setting in the app definition will be used instead if it is configured.

SimonTheLeg reviewed (changes requested):
Overall looks good to me, my comments are just minor adjustments.

However there is one bigger thing on my mind:
What happens with the Cilium system application if a defaultNamespace is set in the Configuration. 
I would really like if we have a test inside cilium that verifies that cilium is unaffected by this value, since it is a special type of system app.

soer3n committed:
avoid copying values and change comment for new value

* use pointer instead of passing value to the function for generating the app namespace
* dereference the returned pointer in the generated applicationinstallation spec
* rename utils.go to namespace.go and test file accordingly
* modify comment for default app namespace field in kubermaticconfiguration

Signed-off-by: soer3n <srenhenning@googlemail.com>

xrstf commented on pkg/apis/kubermatic/v1/configuration.go:
```diff
@@ -354,6 +356,11 @@ type SystemApplicationsConfiguration struct {
 
+// DefaultApplicationsConfiguration contains configuration for default Applications configuration settings.
+type DefaultApplicationsConfiguration struct {
+	Namespace string `json:"namespace,omitempty"`
```
Please document this field. Especially as it seems to be an EE-only functionality.

xrstf replied:
This has not been resolved from what I can see.

soer3n replied:
Sorry. I forgot. Added now a description for namespace field.

xrstf commented on pkg/controller/seed-controller-manager/default-application-controller/controller.go:
```diff
@@ -334,19 +336,27 @@ func (r *Reconciler) generateApplicationInstallation(application appskubermaticv
+	namespace := application.Name
+	config, err := r.configGetter(ctx)
+	if err != nil {
+		r.log.Warnf("failed to check kubermatic configuration default settings for applications: %w", err)
```
Why is this only a warning?

soer3n replied:
This function is not returning an error. The expectation therefore is a working applicationinstallation object. The fallback for this setting is the name of the application. In my opinion a warning should be tracked to give an admin at least an idea why the default setting cannot be used for the object.

xrstf commented on pkg/controller/seed-controller-manager/default-application-controller/wrappers_ee.go:
```diff
@@ -0,0 +1,30 @@
+//go:build ee
+
+/*
+Copyright 2024 The Kubermatic Kubernetes Platform contributors.
```
2025 ;-)

xrstf commented on pkg/ee/default-application-controller/namespace_test.go:
```diff
@@ -0,0 +1,145 @@
+		t.Run(test.name, func(t *testing.T) {
+			// validate the result
+			if !test.validate(test.application) {
+				t.Fatalf("Test failed for: %v with default appnamespace value %v", test.name, test.defaultApplicationNamespace)
```
`Test failed for: %v` is a redundant prefix, given that the test is using subtests already :-)

soer3n replied:
Changed this to a less redundant message.

soer3n committed:
add ee-only comment to new field + change failed test msg

* modify comment for new field in kubermaticconfiguration to make clear that it is only supported in enterprise edition
* change message for failed tests to be not redundant
* change year in header in new created files

Signed-off-by: soer3n <srenhenning@googlemail.com>

ahmedwaleedmalik commented on docs/zz_generated.kubermaticConfiguration.ce.yaml:
```diff
@@ -381,6 +381,9 @@ spec:
     apiserverReplicas: 2
+    # DefaultApplications contains configuration for default Application settings. This is an ee-only feature.
+    defaultApplications:
+      namespace: ""
```
We should rather keep the encapsulating struct generic and call it `applications` or `applicationSettings`. Since default applications != enforced applications and I don't think we should in the future add a separate block for `enforcedApplications` or just `applications`(for generic apps) if we want to expose additional configurations.

ahmedwaleedmalik commented on pkg/controller/seed-controller-manager/default-application-controller/controller.go:
```diff
@@ -334,19 +336,28 @@ func (r *Reconciler) generateApplicationInstallation(application appskubermaticv
+			namespace = config.Spec.UserCluster.DefaultApplications.Namespace
+		}
+	}
+
```
I like the idea here but this can be problematic. Imagine the following scenario:
1. You have multiple defaulted/enforced applications in your user clusters.
2. We change `config.Spec.UserCluster.DefaultApplications.Namespace` to `apps`.
3. We'll end up with duplicate applications in both the `application.name` and `config.Spec.UserCluster.DefaultApplications.Namespace`.

This situation needs to be avoided.

soer3n replied:
Added logic and test to avoid duplicates.

ahmedwaleedmalik commented on pkg/ee/default-application-controller/namespace.go:
```diff
@@ -0,0 +1,39 @@
+		Name:   application.Name,
+		Create: true,
+	}
+}
```
Why is this an EE only feature? Having a default namespace here is essential for some applications to work that require a specific namespace name for the applications to function properly. Making this EE only is not something I'm happy about but if this was already discussed and decided I'm fine, but it'd be helpful for me to know the context.

ahmedwaleedmalik replied:
The field https://github.com/kubermatic/kubermatic/blob/main/pkg/apis/apps.kubermatic/v1/application_definition.go#L223-L225 itself is not marked as EE anyways.

ahmedwaleedmalik replied:
As discussed internally, this will be generally available and not an EE only feature.

ahmedwaleedmalik reviewed (changes requested):
The usage of `defaultNamespace` is quite confusing for me. I don't see any implementation changes in the application controller for [defaultNamespace](https://github.com/kubermatic/kubermatic/blob/main/pkg/apis/apps.kubermatic/v1/application_definition.go#L223-L225) whereas that's the most crucial part and that's why we are introducing this whole field.

If I create a new ApplicationInstallation via CRD, and leave the `spec.Namespace` empty then I'd expect the `defaultNamespace`, if it exists to simply be mutated into the `ApplicationInstallation.spec.namespace`(preferred) so that the `defaultNamespace` is reconciled against https://github.com/kubermatic/kubermatic/blob/main/pkg/applications/installer.go#L111 simply.

Otherwise you are only relying on the KKP API/UI to create the namespace for you which is not ideal.

soer3n committed:
add namespace mutator + ensure appinstall namespace immutability

* move applicationinstallation generation into ensureApplicationInstallation func
** this is to ensure that an applicationinstallation and the related helm release is only installed once
** move getAppNamespace() func from ee to controller because the feature should be for ce and ee
* add applicationinstallation mutation for app namespace when a default is configured and no is specified

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
change field name to application to make it more generic in kkp configuration

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
change client to seed instead of user cluster for mutation webhook

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
add test to check that there is no app duplication when default namespace changed

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
improve scenario description

soer3n committed:
add description for newly introduced field

Signed-off-by: soer3n <srenhenning@googlemail.com>

xrstf commented on pkg/controller/seed-controller-manager/default-application-controller/controller.go:
```diff
@@ -231,64 +235,89 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *kubermaticv1.Cluste
+	// we do this by field selector to avoid to deploy an application twice in different namespaces by mistake
+	existingApplicationList := &appskubermaticv1.ApplicationInstallationList{}
+	if err := userClusterClient.List(ctx, existingApplicationList); err != nil {
+		return fmt.Errorf("failed to fetch existing application: %w", err)
```
failed _to list applications_

not "fetch a single application"

soer3n replied:
Changed it to the suggested wording.

xrstf commented on pkg/controller/seed-controller-manager/default-application-controller/controller.go:
```diff
@@ -231,64 +235,89 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *kubermaticv1.Cluste
+			appDefaultedEnabled, _ := strconv.ParseBool(exisitingApplication.Annotations[appskubermaticv1.ApplicationDefaultedAnnotation])
+			// if enforced and defaulted we found the existing default application installation
+			if appEnforcedEnabled && appDefaultedEnabled {
+				currentApplicationInstallation = &exisitingApplication
```
Could add a `break` statement to signal that you're looking for a unique item. Or do we expect multiple AppInstalls that could potentially conflict with each other?

soer3n replied:
No, a break is ok here because we only expect one applicationinstallation for a default application. Added that.

xrstf commented on pkg/controller/seed-controller-manager/default-application-controller/controller.go:
```diff
@@ -231,64 +235,89 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *kubermaticv1.Cluste
+	}
+
+	if currentApplicationInstallation != nil {
+		if currentApplicationInstallation.Namespace != namespaceName {
```
This could be a single if statement.

soer3n replied:
Changed it to single line statement.

xrstf commented on pkg/controller/seed-controller-manager/default-application-controller/controller.go:
```diff
@@ -231,64 +235,89 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *kubermaticv1.Cluste
+
+	if currentApplicationInstallation != nil {
+		if currentApplicationInstallation.Namespace != namespaceName {
+			r.log.Warnf("namespace for default application installation %s has changed. Keeping the old one because configuring the namespace is only allowed when creating it.", application.Name)
```
Ideally this would make use of stuctured logging, where the app name is a dedicated field on the log line and the text is a bit reworded.

xrstf replied:
"Keeping the old one" -- will this log line be printed over and over and over again? The "keeping" part makes me think this is a kind of permanent state? If so we should not spam the controller logs with a permanent warning IMHO.

soer3n replied:
I removed now the message completely. The namespace is now set back to the one which was used during creation without any log message. Is that ok for you?

xrstf commented on pkg/controller/seed-controller-manager/default-application-controller/controller.go:
```diff
@@ -231,64 +235,89 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *kubermaticv1.Cluste
 		err := userClusterClient.Create(ctx, namespace)
 		if err != nil {
-			return fmt.Errorf("failed to create namespace: %w", err)
+			return fmt.Errorf("failed to creat namespace: %w", err)
```
_create_

xrstf replied:
Also you could consider using the reconciling framework. For namespaces it's often not super duper useful, but it's consistent with the rest of the codebase.

soer3n replied:
Typo is fixed. But here the reconciling framework cannot be used without adding a namespace reconciler to it.

xrstf commented on pkg/controller/seed-controller-manager/default-application-controller/controller.go:
```diff
@@ -231,64 +235,89 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *kubermaticv1.Cluste
+	if err := userClusterClient.Update(ctx, applicationInstallation); err != nil {
 		return fmt.Errorf("failed to update application installation %q: %w", application.Name, err)
 	}
 	return nil
```
Could this entire thing be replaced with the reconciling framework? Where instead of generating the new AppInstall object and then fiddling with ResourceVersion and UID, we simply say `newAppInstall.Spec = generatedAppInstall.Spec`?

soer3n replied:
Replaced with the reconciling framework. Although there was a comment that this framework does not make so much sense here because application installations could have different namespaces and the namespace needs to be passed to the reconciler explicitly, so that the reconciler needs to be called for each generated application installation.

xrstf commented on pkg/controller/seed-controller-manager/default-application-controller/controller.go:
```diff
@@ -362,7 +388,22 @@ func (r *Reconciler) generateApplicationInstallation(application appskubermaticv
+	config, err := r.configGetter(ctx)
+	if err != nil {
+		// This is a non-critical error and we can still continue by using the application name as the namespace
+		r.log.Debugf("failed to check kubermatic configuration default settings for applications: %w", err)
```
This should be an error. Otherwise we create potentially broken and misplaced objects. The config must exist for KKP to work and it will also be in the runtime cache already.

soer3n replied:
Here is now an error returned instead of a debug message.

xrstf commented on pkg/webhook/application/applicationinstallation/mutation/mutation.go:
```diff
@@ -90,3 +99,17 @@ func (h *AdmissionHandler) Handle(ctx context.Context, req webhook.AdmissionRequ
+func mutateAppNamespace(ctx context.Context, seedClient ctrlruntimeclient.Client, applicationInstallation *appskubermaticv1.ApplicationInstallation) error {
+	applicationDefinition := appskubermaticv1.ApplicationDefinition{}
+	if err := seedClient.Get(ctx, types.NamespacedName{Name: applicationInstallation.Spec.ApplicationRef.Name}, &applicationDefinition); err != nil {
+		return fmt.Errorf("error on fetching application definition for mutating appinstallation namespace. %w", err)
```
😬 I can see this causing problems in the future. I have been down this road of trying too much in a webhook. It's a compelling thought: "Let's use webhooks to bring some sequentiality into this eventually consistent system, like a database trigger", but it will not work as you expect. 

Like, I tried to validate a Project object whenever a Cluster object is created and just ran into caching issues (if the Project and Cluster are part of the same YAML file, the cache would not find the Project when it was supposed to validate the Cluster object).

Also, it's possible for your webhook to be disabled or defunct when an AppInstall is meant to be created. It's also possible for objects to be created in other orders, and all Kubernetes users kind of expect to be allowed to create their objects in any order they like and have it be eventually consistent. Adding a webhook to kind of force-sequentialize operations is going against the Kubernetes mantra.

The flipside is that your controller code cannot _rely_ on the webhook's magic. And if you cannot rely on it, your controller code has to deal with all eventualities anyway. And so the webhook doesn't really help that much.

Personally I would recommend never writing webhook logic that relies on _other_ objects in the cluster. It's brittle and if those other objects change, you will not notice and update your own objects (if needed). Usually defaulting logic like this, where a field depends on other objects, should be done at runtime during reconciliation.

Is it possible for this feature to simply accept empty fields in the AppInstall and then handle that during reconciliation?

soer3n replied:
Ok, i can see the problem. I moved the defaulting to the reconcile loop instead. When a default namespace is configured in the app definition and no is set in the application installation the default from the app definition will be used. The app namespace is immutable after creation.

xrstf commented on pkg/controller/seed-controller-manager/default-application-controller/namespace_test.go:
```diff
@@ -0,0 +1,64 @@
+			}
+		})
+	}
+}
```
Let's discuss this for a moment. What is your intention here? You started with table-based tests, but then each subtest has its own validation routine, which is often a code smell for badly designed table tests. Especially if, like in this case, you do not even have shared, common setup logic for each subtest.

Did you plan ahead for much more elaborated tests? I would wager that's maybe a bit overeager and we might never actually need that. What about this simpler version?

```go
func TestGetAppNamespace(t *testing.T) {
	testCases := []struct {
		name        string
		application *appskubermaticv1.ApplicationDefinition
		expected    string
	}{
		{
			name:        "scenario 1: application namespace should be set to default value when a default value is configured",
			application: genApplicationDefinition("applicationName", "namespace", "v1.0.0", "", true, false, defaultValue, nil, &applicationNamespace),
			expected:    applicationNamespace.Name,
		},
		{
			name:        "scenario 2: application namespace should be set to application name when no default value is configured",
			application: genApplicationDefinition("applicationName", "namespace", "v1.0.0", "", true, false, defaultValue, nil, nil),
			expected:    application.Name,
		},
	}

	for _, test := range testCases {
		t.Run(test.name, func(t *testing.T) {
			namespace := getAppNamespace(test.application)
			if namespace != test.expected {
				t.Fatalf("Expected namespace %q, got %q.", test.expected, namespace)
			}
		})
	}
}
```

This actually highlights the common parts and the test expectation. Looking closer, one could even put the `genApplicationDefinition` into the loop below:

```go
func TestGetAppNamespace(t *testing.T) {
	testCases := []struct {
		name        string
		appNamespace *appskubermaticv1.AppNamespaceSpec
		expected    string
	}{
		{
			name:         "scenario 1: application namespace should be set to application name when no default value is configured",
			appNamespace: nil,
			expected:     application.Name,
		},
		{
			name:         "scenario 2: application namespace should be set to default value when a default value is configured",
			appNamespace: applicationNamespace,
			expected:     applicationNamespace.Name,
		},
	}

	for _, test := range testCases {
		t.Run(test.name, func(t *testing.T) {
			application := genApplicationDefinition("applicationName", "namespace", "v1.0.0", "", true, false, defaultValue, nil, test.appNamespace),
			namespace := getAppNamespace(test.application)

			if namespace != test.expected {
				t.Fatalf("Expected namespace %q, got %q.", test.expected, namespace)
			}
		})
	}
}
```

Now the test really shows only the interesting bits per subtest. It also gives a better error output without the redundant subtest name in it.

soer3n replied:
Thank you for the hint. I changed that and used the pattern also for new added tests.

soer3n committed:
change ensuring appinstallation to reconciler framework

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
modify table test structure + move defaulting to reconcile loop

* remove validate func from unit tests and check instead directly inside the test table loop
* move defaulting if no app namespace is configured in the application installation at the start of reconcile instead of webhook mutation

Signed-off-by: soer3n <srenhenning@googlemail.com>

xrstf commented on pkg/controller/user-cluster-controller-manager/application-installation-controller/controller.go:
```diff
@@ -214,6 +218,20 @@ func (r *reconciler) getApplicationVersion(appInstallation *appskubermaticv1.App
+func setAppNamespace(ctx context.Context, seedClient ctrlruntimeclient.Client, applicationInstallation *appskubermaticv1.ApplicationInstallation) error {
+	applicationDefinition := appskubermaticv1.ApplicationDefinition{}
+	if err := seedClient.Get(ctx, types.NamespacedName{Name: applicationInstallation.Spec.ApplicationRef.Name}, &applicationDefinition); err != nil {
+		return fmt.Errorf("error on fetching application definition for setting appinstallation namespace. %w", err)
```
`: %w` at the end for better error formatting. Also you can get rid of "for setting appinstallation namespace", because the calling function will prepend a very similar text to the error message already.

xrstf commented on pkg/controller/seed-controller-manager/default-application-controller/namespace_test.go:
```diff
@@ -0,0 +1,63 @@
+			// validate the result
+			if namespace.Name != test.expectedNamespace {
+				t.Fatalf("Validation failed for %v", test.name)
+			}
```
Let's have some meaningful error messages that don't repeat the subtest name. Maybe `Expected namespace %q, got %q instead.` or something.

soer3n committed:
make error messages in controller and testing a bit clearer

Signed-off-by: soer3n <srenhenning@googlemail.com>

xrstf reviewed (approved):
/approve
